python/cuml/metrics/confusion_matrix.py

```diff
@@ -0,0 +1,114 @@
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+import numpy as np
+import cupy as cp
+
+from cuml.utils import input_to_cuml_array
+from cuml.utils.memory_utils import with_cupy_rmm
+from cuml.metrics.utils import sorted_unique_labels
+from cuml.prims.label import make_monotonic
+
+
+@with_cupy_rmm
+def confusion_matrix(y_true, y_pred,
+                     labels=None,
+                     sample_weight=None,
+                     normalize=None):
+    """Compute confusion matrix to evaluate the accuracy of a classification.
+
+    Parameters
+    ----------
+    y_true : array-like (device or host) shape = (n_samples,)
+        or (n_samples, n_outputs)
+        Ground truth (correct) target values.
+    y_pred : array-like (device or host) shape = (n_samples,)
+        or (n_samples, n_outputs)
+        Estimated target values.
+    labels : array-like (device or host) shape = (n_classes,), optional
+        List of labels to index the matrix. This may be used to reorder or
+        select a subset of labels. If None is given, those that appear at least
+        once in y_true or y_pred are used in sorted order.
+    sample_weight : array-like (device or host) shape = (n_samples,), optional
+        Sample weights.
+    normalize : string in [‘true’, ‘pred’, ‘all’]
+        Normalizes confusion matrix over the true (rows), predicted (columns)
+        conditions or all the population. If None, confusion matrix will not be
+        normalized.
+
+    Returns
+    -------
+    C : array-like (device or host) shape = (n_classes, n_classes)
+        Confusion matrix.
+    """
+    y_true, n_rows, n_cols, dtype = \
+        input_to_cuml_array(y_true, check_dtype=[cp.int32, cp.int64])
+    y_true = y_true.to_output('cupy')
```

Comment:
you don't need to do this conversion, CumlArray supports `__cuda_array_interface__` so you should just be able to use it as a parameter in the functions of CuPy below

Reply:
I think it is needed for `make_monotonic` later because it uses a cuPy RawKernel. Also later I use comparison operator `y_pred < n_labels` and the indexing operator `sample_weight[ind]`. I know could replace them with `cp.less` and `__getitem__` but it makes the code harder to read in my opinion.

Reply:
And from my tests `.to_output('cupy')` doesn't create a copy unless it has to (CPU to GPU for instance).

Reply:
I believe you can use CumlArray as an input for  `make_monotonic` function. Also I feel it is better to avoid any possibility of unnecessary memory usage as we do not know what type of input the user will be providing to this metric.

Reply:
Nevermind, I tried again today and you guys are correct, I must have missed something the first time. I removed the unnecessary `.to_output('cupy')`.